File: chapter7/lambda/lambda_examples.cpp
// JU26625

// Anonymous functions & lambda expressions

// The core syntax difference:
// C++ added lambda expressions / anonymous functions in C++11 (2011) with full closure capture semantics.
// C++ lambdas are full statement blocks ([capture](params) { body }); more powerful than Python lambdas at the cost of verbosity.

// Import libraries
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;


// Example 1: double a number
int double_int(int x) {          // fix: missing type "int" before x
    return 2 * x;
}

auto f = [](int x) {
    return x * 2;
};                               // fix: missing semicolon


// Example 2: add two numbers
int add(int x, int y) {
    return x + y;
}

auto add_lambda = [](int x, int y) {
    return x + y;
};


// Example 3: filter even numbers
bool is_even(int x) {
    return x % 2 == 0;
}


// Main function
int main() {

    // Example 1: double a number ..................................................................
    cout << "\nDouble a number" << endl;
    cout << "With double_int(x): " << double_int(5) << endl;
    cout << "With lambda f(x): " << f(5) << endl;         // fix: missing << before f(5)
    // Normal function: human-readable name
    cout << typeid(double_int).name() << endl; // "FiiE" or similar (mangled)

    // Example 2: add two numbers ..................................................................
    cout << "\nAdd two numbers" << endl;
    cout << "With add(x, y): " << add(2, 3) << endl;
    cout << "With lambda add(x, y): " << add_lambda(2, 3) << endl;
    // Lambda: compiler-generated unique type, unreadable
    cout << typeid(f).name() << endl; // "Z4mainEUliE_" or similar
    
    // Example 3: filter even numbers ..............................................................
    cout << "\nFilter even numbers" << endl;
    vector<int> numbers = {1, 2, 3, 4, 5, 6, 7, 8, 9};
    vector<int> evens;

    // With a normal function
    copy_if(numbers.begin(), numbers.end(), back_inserter(evens), is_even);
    cout << "With is_even(x): ";
    for (int n : evens) cout << n << " ";
    cout << endl;

    // With a lambda expression
    evens.clear();
    copy_if(numbers.begin(), numbers.end(), back_inserter(evens),
            [](int x) { return x % 2 == 0; });
    cout << "With lambda(x): ";
    for (int n : evens) cout << n << " ";
    cout << endl;

    return 0;

}

File: chapter7/lambda/lambda_examples.py
# JU26625

# Anonymous functions & lambda expressions

# The core syntax difference:
# Python added lambda expressions in 1994 (v1.0) as anonymous single-expression functions
# Python lambdas are expression-only (lambda params: expr); simple than C++ lambdas but less control



# Example 1: double a number ......................................................................
print("\nDouble a number")
# With a normal function
def double(x):
    return 2 * x
print(f"With double(x) function: {double(5)}")
print(double.__name__)

# With a lambda expression
f = lambda x: 2 * x
print(f"With lambda f(x) function: {f(5)}")
print(f.__name__)



# Example 2: add two numbers ......................................................................

# With a normal function
def add(x, y):
    return x + y
print(f"With add(x, y) function: {add(2, 3)}")
print(add.__name__)

# With a lambda expression
f = lambda x, y: x + y
print(f"With lambda add(x, y) function: {f(2, 3)}")
print(f.__name__)



# Example 3: filter even numbers ..................................................................

# With a normal function
def is_even(x):
    return x % 2 == 0
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9]
evens = list(filter(is_even, numbers))
print(f"With is_even(x) function: {evens}")

# With a lambda expression
evens = list(filter(lambda x: x % 2 == 0, numbers))
print(f"With lambda(x) function: {evens}")
